ReciprocalSearch improvements

TW80000 opened the pull request:
## Changes
* Update UI
* Property filters are now hidden behind a checkbox by default
* Fix serialization for Object properties that get set to Date values (fixes dates being used in binary mlangs like Gt and Lt, which are used in DateSearchView)
* Clear button still clears each search view, but now also collapses all property search views

## New features
* Add a search view for Date properties
* Add a search view for Currency properties

## New UI
[Original designs](https://app.zeplin.io/project/58ed17e322a7ba2e76e55cde/screen/5a0f3cb00c6b67ef13d2f8b8)

<img width="315" alt="screen shot 2018-08-01 at 1 54 49 pm" src="https://user-images.githubusercontent.com/4259165/43539298-81b81148-9592-11e8-94bb-6dcd23d8614e.png">

Commit:
Improve code style in ReciprocalSearch.js

Commit:
Make ReciprocalSearch a View

Commit:
Improve code formatting in GroupBySearchView.js

Commit:
Fix bug in GroupBySearchView

The bug is that when hardData is set to a falsey value, it isn't saved.
This commit fixes that by checking that hardData is undefined, not just
falsey, which could be a valid value for it to have.

For example: lazy = false would set hardData to false, which is a valid
value for it to be. However, the weak check to see if it was set would
treat it as unset.

Commit:
Add ability to toggle properties in ReciprocalSearch

Commit:
Update TextSearchView style in ReciprocalSearch

Commit:
Linting

Commit:
Add step attribute to FloatView

Commit:
Create CurrencySearchView

Commit:
Remove console.log

Commit:
Code style improvements

Commit:
Style selects in filter panel

Commit:
Create date parser for Object properties

Commit:
Create search view for Dates

TW80000 commented:
Closing because it's not quite ready yet.

Commit:
Fix DateSearchView CSS

Commit:
Finish documentation

TW80000 commented:
@zahnno Please review. I will be applying this to ExpensesView and SalesView in a NANOPAY PR later as per https://github.com/nanoPayinc/NANOPAY/issues/2401.

@kgrgreer Please review, particularly changes to Date (de)serialization.

@jingyilai0809 Please review, particularly the UI.

jingyilai0809 commented:
My notes
1. will need to think about the order of the filter items
2. can we change the background color to match the design? also the checkbox @TW80000 
![image](https://user-images.githubusercontent.com/27588387/43540578-22924d2e-9596-11e8-859f-0c88a93bf3e1.png)
3. Font family should match the design (for next time as it is a lot of work to change according to Troy)

TW80000 commented:
To expand on @jingyilai0809's third point, currently a monospace font is being used in GroupBySearchView because we're currently using a bunch of non-breaking spaces to separate the name of an entry in the list from its count. Monospace fonts have fixed-width characters so they all line up. If we were to switch to a non-monospace font in that view - like the design calls for - the alignment would break. The way we do the formatting should be improved.

Commit:
Merge commit '96fedd3b8' into filter-panel

kgrgreer commented:
Can be done in a future PR, but the formatting under 'Status' doesn't match the rest of the views.

kgrgreer commented on src/foam/lib/json/ObjectDateParser.java:
```diff
@@ -0,0 +1,101 @@
+public class ObjectDateParser
+  extends ProxyParser
+{
+  protected static ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
```
Is this StringBuilder used?

Commit:
Move input style to higher UI component

Commit:
Fix outline on ReciprocalSearch inputs

Commit:
Add general MQL query to ReciprocalSearch

Commit:
Remove unnecessary property assignment

SearchViewWrapper doesn't have a predicate property. That assignment was
a remnant from previous work while I was still writing SearchViewWrapper.

Commit:
Remove unused code

TW80000 commented:
## Update

* General MQL search was added
* Input outline style changed from default to blue border

<img width="318" alt="screen shot 2018-08-03 at 1 48 15 pm" src="https://user-images.githubusercontent.com/4259165/43657489-01f4af64-9724-11e8-81b2-98cc776fadf6.png">

## TODO

* Fix font used in GroupBy search views
* Fix status styling - @kgrgreer What do you mean by this?
* ~~Fix background colour~~
* ~~Fix checkbox colour~~

Commit:
Prevent checkboxes from being visually focused

Commit:
Fix background style in SearchViewWrapper

TW80000 commented:
## Update

* Fixed background colour

<img width="318" alt="screen shot 2018-08-03 at 2 09 08 pm" src="https://user-images.githubusercontent.com/4259165/43658439-e93b727a-9726-11e8-9c8f-ec6e219a98b1.png">

kgrgreer commented on src/foam/u2/FloatView.js:
```diff
@@ -51,22 +64,22 @@ foam.CLASS({
       this.precision = p.precision;
-      this.min       = p.min;
-      this.max       = p.max;
+      this.min = p.min;
```
Why messing up alignment?

TW80000 replied:
[Horizontal alignment is discouraged in the Google JavaScript Styleguide](https://google.github.io/styleguide/jsguide.html#formatting-horizontal-alignment). Our linter settings also flag it as an error.

kgrgreer replied:
I've just updated the FOAM guide to officially reflect what has been the de-facto standard all along.

kgrgreer commented on src/foam/u2/FloatView.js:
```diff
@@ -51,22 +64,22 @@ foam.CLASS({
       val = val.toFixed(this.precision);
       var i = val.length - 1;
-      for ( ; i > 0 && val.charAt(i) === '0' ; i-- ) {}
+      for ( ; i > 0 && val.charAt(i) === '0'; i -- ) {}
```
The space between i and -- doesn't comply with the google/foam style guideline.

TW80000 replied:
You're right. Since we enforce a space between `!` and its argument as part of our rule exceptions, our linter flags all unary operators without a space between them and their arguments. I'll revert these spots back and have a quick look to see if the linter settings can be more fine-grained.

kgrgreer commented on src/foam/u2/search/CurrencySearchView.js:
```diff
@@ -0,0 +1,152 @@
+          to filter the dao being displayed in the view.`,
+      expression: function(qualifier, amount) {
+        amount = typeof amount === 'number' ?
+            Math.floor(amount * 100) :
```
If this is the only thing that distinguishes this from a regular number searchView, couldn't this whole class be made reusable and have just the line abstracted out to a template method in a very small sub-clas?

TW80000 replied:
Yeah, that makes sense. I'll split this into NumberSearchView and CurrencySearchView.

kgrgreer commented on src/foam/u2/search/DateSearchView.js:
```diff
@@ -0,0 +1,127 @@
+          SearchManager. The SearchManager will read this predicate and use it
+          to filter the dao being displayed in the view.`,
+      expression: function(qualifier, date) {
+        if ( qualifier.length > 0 && date && ! isNaN(date.valueOf()) ) {
```
If Gt it should Gt end-of-day. If Lt it should be Lt end-of-day. 
As mentioned in demos, Between would also be good.
Other options might include: 'Yesterday', 'Today', 'Tomorrow', 'This Week', 'This Month', 'Next Month', etc.

TW80000 replied:
This is more or less blocked by the big date/time refactoring that @mcarcaso proposed. I could implement this before the refactor, but then I'd need to go back and change it again afterwards.

kgrgreer replied:
Okay, let's just do in a followup PR.

kgrgreer commented on src/foam/u2/search/DateSearchView.js:
```diff
@@ -0,0 +1,127 @@
+    }
+  ],
+
+  css: `
```
I prefer to put css near the top with constants and messages since it's more of a class/static kind of thing. But your choice.

TW80000 replied:
I chose to put it at the bottom because CSS usually takes up a lot of space and I rarely need to look at it after it's written the first time. It's at the bottom to avoid having to scroll past it every time I want to add a require or check a property and then go back to the initE function. I'll just leave it at the bottom if it's up to me.

kgrgreer replied:
ok

kgrgreer commented on src/foam/u2/view/SearchViewWrapper.js:
```diff
@@ -0,0 +1,120 @@
+    predicate.`,
+
+  requires: [
+    'foam.mlang.predicate.True',
```
No trailing ,

kgrgreer commented on src/foam/u2/view/SearchViewWrapper.js:
```diff
@@ -0,0 +1,120 @@
+      name: 'active',
+      documentation: `Tracks whether the property is being used as part of the
+        filter criteria or not.`,
+      value: false
```
value: false is the default, so this line is redundant

kgrgreer commented on src/foam/u2/view/SearchViewWrapper.js:
```diff
@@ -0,0 +1,120 @@
+      this.checkbox.data$.sub(this.checkboxChanged);
+    },
+
+    function toggleActive() {
```
This could just be put in the listener.

kgrgreer commented:
Good job. Added some comments.

Commit:
Code style changes for peer review

Commit:
Create search views for all numeric types

Commit:
Fix code style

Commit:
Update checkbox colour

TW80000 commented:
## Update

* Changed checkbox colour

<img width="315" alt="screen shot 2018-08-03 at 3 52 43 pm" src="https://user-images.githubusercontent.com/4259165/43662859-5300fa50-9735-11e8-970b-a0338194b0e6.png">

jingyilai0809 commented:
@TW80000 
I really appreciate the to-do list, a title and a screenshot to review the design.
It makes things a lot easier to understand and provide lots of clarity.
Thank you so much for doing that!!!! 
@kgrgreer 
It will be great if we can adopt the same approach for every PR which requires a designer's review.

**Feedback:**
1. Checkbox is good
2. Background is good
3. The search field styling is matched

Good job! Thank you!

kgrgreer commented on src/foam/u2/search/FilterController.js:
```diff
@@ -505,3 +505,123 @@ foam.CLASS({
+
+foam.CLASS({
+  package: 'foam.u2.search',
+  name: 'DoubleRefinement',
```
Double extends Float so just refining Float is sufficient.

kgrgreer commented on src/foam/u2/search/IntegerSearchView.js:
```diff
@@ -0,0 +1,152 @@
+foam.CLASS({
+  package: 'foam.u2.search',
+  name: 'IntegerSearchView',
+  extends: 'foam.u2.View',
```
If you extend Controller instead of View then you will get the exports: 'as data' for free. Generally, a View is like an MVC View which is bound to some data, but in this case the GUI is its own data, which makes it more of a Controller.

Commit:
Remove redundant refinement

Since Double extends Float, refining Double is redundant.

Commit:
Fix bug in IntegerSearchView

There was code left over from CurrencySearchView that multiplies the
number by 100. That needs to be done for currencies because we transfer
and store currency values as integer amounts of cents, but we input them
like floats (ie: with a decimal point).

This commit removes that code and fixes the issue for normal,
non-currency search view inputs.

Commit:
Make SearchViews extend Controller instead of View

That way they don't have to explicitly state that they export themselves
"as data".

Commit:
Lint Select.js

Commit:
Merge branch 'master' into filter-panel

Commit:
Fix broken select element height

I forgot to add a caret here before which limits the scope of the CSS.
That caused all select elements to be a fixed height.

Commit:
Make size of select in ChoiceView dynamic

Instead of ChoiceView being fixed at size 10, make it dynamically update
to match the size of the list of choices.

Commit:
Revert "Make size of select in ChoiceView dynamic"

This reverts commit cb5b6316c45d03aa16908e1118365b60fc5b1d59.

Reverting because it broke other places where it doesn't make sense to
see a full list of options.

Commit:
Add option to make ChoiceView's size dynamic

Commit:
Make GroupBySearchView use dynamic size ChoiceView

Was fixed at size 10 before, but that fixed size made lists with less
than 10 entries take up more space than necessary. This is a more
flexible solution that makes the size of the list dynamic, while still
letting you set a maximum size so that the list doesn't take up more
space than you want it to.
